### PR TITLE
282: Fix modular-spacing mixins and variables

### PR DESCRIPTION
**Pull request by yvonnetangsu**
# READY

# Summary
- Fix modular-spacing related mixins, functions and variables so the values match the Figma file

# Needed By (Date)
- N/A

# Urgency
- Not urgent

# Steps to Test
- pull this branch and run grunt styleguide
- check that the variables in variables/core/_modular-spacing.scss are correct and the mixins/functions under modular-scale (the spacing related ones) work as intended (Figma documentation under Stanford Design Library -> Visual Guidelines -> Spacing)
-  _card.scss, _card--horizontal.scss now uses modular-padding mixin instead of old $modular-spacing variables - please check if that's ok
- core/scss/core/flex-grid/_flex-grid.scss now uses $grid-gutter values instead of old $modular-spacing variables - please check if that's ok

# Affected Projects or Products
- Decanter

# Associated Issues and/or People
- Affected components - cards, flex-grid (ones that use modular-spacing, modular-margins etc @sherakama @josephgknox )
- Note: Modular spacing and modular typography use totally different system to calculate the steps, even though they share the same base value at each breakpoint.
- Modular typography follows the Bourbon modular-scale mixin formula with a 1.25 scale ratio between each step, ie -
modular-scale(0) = $base-value
modular-scale(1) = $base-value * 1.25
modular-scale(2) = $base-value * 1.25 * 1.25 etc
- Modular spacing does NOT use the above system. There is no fix scale ratio value. Instead, the scale is manually adjusted by John and Kerri because they needed more granular control over the spacing values. This scale is documented in Figma.

Also,

- Modular spacing, screen margins, gutters are unrelated

### CODE REVIEW

**Comment by yvonnetangsu** [comment]
screen shot 2019-01-25 at 11 00 46 am

**Comment by sherakama** [comment]
@yvonnetangsu Thanks for this work. The render looks good. I will perform the code review next. To get this sorted out for myself and to make sure we don't overlap in namespacing the different systems I have started the documentation on the issue ticket: https://github.com/SU-SWS/decanter/issues/288

Please take a look at that.

File: core/scss/utilities/functions/modular-scale/_get-modular-base.scss [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+// get-modular-base($breakpoint)
+//
+// Get a value out of the modular-spacing-base map
+// Returns the pixel value of the $modular-spacing-base variable at a breakpoint.
+//
+// $breakpoint - One of `xs`, `sm`, `md`, `lg`, `xl`, `2xl`
+//
+// Style guide: Functions.ModularScale.get-modular-base
+//
+@function get-modular-base($breakpoint: 'xs') {
```
Comment (sherakama) [inline core/scss/utilities/functions/modular-scale/_get-modular-base.scss]: perhaps we call this `get-modular-spacing-base` for a slightly better namespace.

File: core/scss/utilities/functions/modular-scale/_modular-spacing.scss [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+// Returns the modular spacing value at a step(integer) on the scale at a certain breakpoint.
+// The modular spacing value is calculated by multiplying the modular spacing base value at the requested breakpoint and the multiplier at the requested step of the scale.
+// The default value for $step is 0, which returns a multiplier value of 1, i.e., modular-spacing(0) = $modular-spacing-base at that breakpoint
+//
+// $step - Integers from -4 to 5 including 0
+// $breakpoint - One of `xs`, `sm`, `md`, `lg`, `xl`, `2xl`
+//
+// Style guide: Functions.ModularScale.modular-spacing
+//
+@function modular-spacing($step: 0, $breakpoint: 'xl') {
```
Comment (sherakama) [inline core/scss/utilities/functions/modular-scale/_modular-spacing.scss]: This is great.

File: core/scss/utilities/functions/modular-scale/_get-modular-spacing-multiplier.scss [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+//
+// Get a value out of the modular-spacing-multiplier map.
+// Returns the value of the $modular-spacing-multiplier variable at a step(integer) on the scale.
+// The default for $step is 0, which returns a multiplier value of 1.
+//
+// $step - Integers from -4 to 5 including 0
+//
+// Style guide: Functions.ModularScale.get-modular-spacing-multiplier
+//
+@function get-modular-spacing-multiplier($step: 0) {
```
Comment (sherakama) [inline core/scss/utilities/functions/modular-scale/_get-modular-spacing-multiplier.scss]: I think this needs some error handling. What if I am a bad user and send in something that isn't -4 through 5?

File: core/scss/utilities/mixins/modular-scale/_modular-margin-bottom.scss [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+//
+// @modular-margin-bottom
+//
+// A mixin for adding modular margin-bottom to an element.
+//
+// $step - Integers from -4 to 5 including 0. Default for $step is 0 which returns the modular spacing base value.
+//
+// Style guide: Mixins.Modular.modular-margin-bottom
+//
+@mixin modular-margin-bottom($step: 0, $grid: $grid-media) {
```
Comment (sherakama) [inline core/scss/utilities/mixins/modular-scale/_modular-margin-bottom.scss]: I get why you have created this function but it seems overly specific and in the previous rendition of the modular functions I wrote something similar but I was not overly fond of it. 

Do you think we could refactor these modular-margin-vertical modular-margin-bottom to perhaps pass in the css property we would like to have modular value on?

eg: 
```
function modular-value-property($step, $property, $unit) {
... code ...
}
```

usage:
```
p {
  @include modular-value-property(1, 'margin-bottom', 'px')
}
```

File: core/scss/utilities/mixins/modular-scale/_modular-margin-horizontal.scss [hunk 5]
```diff
@@ -3,31 +3,16 @@
...
-    @include margin(0 map-get($modular-spacing, '2xl') * $scale);
-  }
-
-  @include grid-media('xl') {
-    @include margin(0 map-get($modular-spacing, 'xl') * $scale);
-  }
-
-  @include grid-media('lg') {
-    @include margin(0 map-get($modular-spacing, 'lg') * $scale);
+@mixin modular-margin-horizontal($step: 0, $grid: $grid-media) {
```
Comment (sherakama) [inline core/scss/utilities/mixins/modular-scale/_modular-margin-horizontal.scss]: see comment on modular-margin-bottom

File: core/scss/utilities/mixins/modular-scale/_modular-margin-left.scss [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+//
+// @modular-margin-left
+//
+// A mixin for adding modular margin-left to an element.
+//
+// $step - Integers from -4 to 5 including 0. Default for $step is 0 which returns the modular spacing base value.
+//
+// Style guide: Mixins.Modular.modular-margin-left
+//
+@mixin modular-margin-left($step: 0, $grid: $grid-media) {
```
Comment (sherakama) [inline core/scss/utilities/mixins/modular-scale/_modular-margin-left.scss]: see comment on modular-margin-bottom

File: core/scss/utilities/mixins/modular-scale/_modular-margin-right.scss [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+//
+// @modular-margin-right
+//
+// A mixin for adding modular margin-right to an element.
+//
+// $step - Integers from -4 to 5 including 0. Default for $step is 0 which returns the modular spacing base value.
+//
+// Style guide: Mixins.Modular.modular-margin-right
+//
+@mixin modular-margin-right($step: 0, $grid: $grid-media) {
```
Comment (sherakama) [inline core/scss/utilities/mixins/modular-scale/_modular-margin-right.scss]: see comment on modular-margin-bottom

File: core/scss/utilities/mixins/modular-scale/_modular-margin-top.scss [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+//
+// @modular-margin-top
+//
+// A mixin for adding modular margin-top to an element.
+//
+// $step - Integers from -4 to 5 including 0. Default for $step is 0 which returns the modular spacing base value.
+//
+// Style guide: Mixins.Modular.modular-margin-top
+//
+@mixin modular-margin-top($step: 0, $grid: $grid-media) {
```
Comment (sherakama) [inline core/scss/utilities/mixins/modular-scale/_modular-margin-top.scss]: see comment on modular-margin-bottom

File: core/scss/utilities/mixins/modular-scale/_modular-margin-vertical.scss [hunk 9]
```diff
@@ -3,31 +3,16 @@
...
-    @include margin(map-get($modular-spacing, '2xl') * $scale 0);
-  }
-
-  @include grid-media('xl') {
-    @include margin(map-get($modular-spacing, 'xl') * $scale 0);
-  }
-
-  @include grid-media('lg') {
-    @include margin(map-get($modular-spacing, 'lg') * $scale 0);
+@mixin modular-margin-vertical($step: 0, $grid: $grid-media) {
```
Comment (sherakama) [inline core/scss/utilities/mixins/modular-scale/_modular-margin-vertical.scss]: see comment on modular-margin-bottom

File: core/scss/utilities/mixins/modular-scale/_modular-margin.scss [hunk 10]
```diff
@@ -3,31 +3,16 @@
...
-    @include margin(map-get($modular-spacing, '2xl') * $scale);
-  }
-
-  @include grid-media('xl') {
-    @include margin(map-get($modular-spacing, 'xl') * $scale);
-  }
-
-  @include grid-media('lg') {
-    @include margin(map-get($modular-spacing, 'lg') * $scale);
+@mixin modular-margin($step: 0, $grid: $grid-media) {
```
Comment (sherakama) [inline core/scss/utilities/mixins/modular-scale/_modular-margin.scss]: See comment on modular-margin-bottom

File: core/scss/utilities/mixins/modular-scale/_modular-padding-bottom.scss [hunk 11]
```diff
@@ -0,0 +1,18 @@
...
+//
+// @modular-padding-bottom
+//
+// A mixin for adding modular padding-bottom to an element.
+//
+// $step - Integers from -4 to 5 including 0. Default for $step is 0 which returns the modular spacing base value.
+//
+// Style guide: Mixins.Modular.modular-padding-bottom
+//
+@mixin modular-padding-bottom($step: 0, $grid: $grid-media) {
```
Comment (sherakama) [inline core/scss/utilities/mixins/modular-scale/_modular-padding-bottom.scss]: see comment on: modular-margin-bottom

File: core/scss/utilities/mixins/modular-scale/_modular-padding-horizontal.scss [hunk 12]
```diff
@@ -0,0 +1,18 @@
...
+//
+// @modular-padding-horizontal
+//
+// A mixin for adding modular left/right padding to an element.
+//
+// $step - Integers from -4 to 5 including 0. Default for $step is 0 which returns the modular spacing base value.
+//
+// Style guide: Mixins.Modular.modular-padding-horizontal
+//
+@mixin modular-padding-horizontal($step: 0, $grid: $grid-media) {
```
Comment (sherakama) [inline core/scss/utilities/mixins/modular-scale/_modular-padding-horizontal.scss]: see comment on modular-margin-bottom

File: core/scss/utilities/mixins/modular-scale/_modular-padding-left.scss [hunk 13]
```diff
@@ -0,0 +1,18 @@
...
+//
+// @modular-padding-left
+//
+// A mixin for adding modular padding-left to an element.
+//
+// $step - Integers from -4 to 5 including 0. Default for $step is 0 which returns the modular spacing base value.
+//
+// Style guide: Mixins.Modular.modular-padding-left
+//
+@mixin modular-padding-left($step: 0, $grid: $grid-media) {
```
Comment (sherakama) [inline core/scss/utilities/mixins/modular-scale/_modular-padding-left.scss]: see comment on: modular-margin-bottom

File: core/scss/utilities/mixins/modular-scale/_modular-padding-right.scss [hunk 14]
```diff
@@ -0,0 +1,18 @@
...
+//
+// @modular-padding-right
+//
+// A mixin for adding modular padding-right to an element.
+//
+// $step - Integers from -4 to 5 including 0. Default for $step is 0 which returns the modular spacing base value.
+//
+// Style guide: Mixins.Modular.modular-padding-right
+//
+@mixin modular-padding-right($step: 0, $grid: $grid-media) {
```
Comment (sherakama) [inline core/scss/utilities/mixins/modular-scale/_modular-padding-right.scss]: see comment on: modular-margin-bottom

File: core/scss/utilities/mixins/modular-scale/_modular-padding-top.scss [hunk 15]
```diff
@@ -0,0 +1,18 @@
...
+//
+// @modular-padding-top
+//
+// A mixin for adding modular padding-top to an element.
+//
+// $step - Integers from -4 to 5 including 0. Default for $step is 0 which returns the modular spacing base value.
+//
+// Style guide: Mixins.Modular.modular-padding-top
+//
+@mixin modular-padding-top($step: 0, $grid: $grid-media) {
```
Comment (sherakama) [inline core/scss/utilities/mixins/modular-scale/_modular-padding-top.scss]: see comment on modular-margin-bottom

File: core/scss/utilities/mixins/modular-scale/_modular-padding-vertical.scss [hunk 16]
```diff
@@ -0,0 +1,18 @@
...
+//
+// @modular-padding-vertical
+//
+// A mixin for adding modular top/bottom padding to an element.
+//
+// $step - Integers from -4 to 5 including 0. Default for $step is 0 which returns the modular spacing base value.
+//
+// Style guide: Mixins.Modular.modular-padding-vertical
+//
+@mixin modular-padding-vertical($step: 0, $grid: $grid-media) {
```
Comment (sherakama) [inline core/scss/utilities/mixins/modular-scale/_modular-padding-vertical.scss]: see comment on modular-margin-bottom

File: core/scss/utilities/mixins/modular-scale/_modular-padding.scss [hunk 17]
```diff
@@ -0,0 +1,18 @@
...
+//
+// @modular-padding
+//
+// A mixin for adding modular padding around an element
+//
+// $step - Integers from -4 to 5 including 0. Default for $step is 0 which returns the modular spacing base value.
+//
+// Style guide: Mixins.Modular.modular-padding
+//
+@mixin modular-padding($step: 0, $grid: $grid-media) {
```
Comment (sherakama) [inline core/scss/utilities/mixins/modular-scale/_modular-padding.scss]: see comment on modular-margin-bottom

**Comment by sherakama** [comment]
This is great. Only one major change and that is in consolidating all of the margin and padding mixins. I think we can refactor those in to one function that spits out what we want.

File: core/scss/utilities/functions/modular-scale/_get-modular-spacing-multiplier.scss (hunk 3 above)
Comment (josephgknox) [reply core/scss/utilities/functions/modular-scale/_get-modular-spacing-multiplier.scss]:
``` 
 // If user passes through non-valid `$step`, return `null`
    @if $step < -4 {
      @return 'null';
    }
    @if $step < 5 {
      @return 'null';
    }
```

Comment (josephgknox) [reply core/scss/utilities/functions/modular-scale/_get-modular-spacing-multiplier.scss]:
```
// If user passes through non-valid `$step`, return `null`
@if $step < -4 {
  @error 'Step`#{$step}` is not a valid modular value';
}
@if $step < 5 {
  @error 'Step`#{$step}`  is not a valid modular value;
}
```

Comment (sherakama) [reply core/scss/utilities/functions/modular-scale/_get-modular-spacing-multiplier.scss]: In this case, you would not want to return null as that could have possible downstream problems and actually make it more difficult to debug. Errors are useful and in this case, you want to throw an error and stop compilation if you get a bad value.

also, you can combine your if statements here with an `or` operator. 

https://stackoverflow.com/questions/17225229/using-multiple-conditions-and-in-sass-if-statement/17226136

Comment (josephgknox) [reply core/scss/utilities/functions/modular-scale/_get-modular-spacing-multiplier.scss]: That makes sense. Also, thanks for the link; super useful operators there.

Would we just want to throw an error then (something closer to the second example above)?

Comment (yvonnetangsu) [reply core/scss/utilities/functions/modular-scale/_get-modular-spacing-multiplier.scss]: Thanks guys! I'll add the error handling.

File: core/scss/utilities/mixins/modular-scale/_modular-margin-bottom.scss (hunk 4 above)
Comment (yvonnetangsu) [reply core/scss/utilities/mixins/modular-scale/_modular-margin-bottom.scss]: Yeah, I was thinking all the mixins were a bit redundant too. Sure, I'll rewrite this part to consolidate. Should we support 'px' unit or just allow 'em'?

Comment (sherakama) [reply core/scss/utilities/mixins/modular-scale/_modular-margin-bottom.scss]: Good question. Let's start with just `em` and do not allow for the unit value in the mixin. We can adjust that later if we need to expand.

**Comment by josephgknox** [comment]
Yvonne is killing it!!

Regard CC: unless it’s something that you feel should be globally changed (even if only for readability; and doesn’t stray away unnecessarily from standards), I would like to clean up the empty space so that CC is happy. Whatcha think?

**Comment by yvonnetangsu** [comment]
Sure, makes sense to keep things consistent. I'll fix the cc complains tomorrow.

**Comment by josephgknox** [comment]
Sweet! Thanks!

File: core/scss/utilities/functions/modular-scale/_calculate-modular-spacing.scss [hunk 18]
```diff
@@ -0,0 +1,23 @@
...
+// The modular spacing value is calculated by multiplying the modular spacing base value at the requested breakpoint and the multiplier at the requested step of the scale.
+//
+// The default value for $step is 0, which returns a multiplier value of 1, i.e., calculate-modular-spacing(0, 'xs') = $modular-spacing-base at XS breakpoint
+//
+// $step - Integers from -4 to 5 including 0
+// $breakpoint - One of `xs`, `sm`, `md`, `lg`, `xl`, `2xl`
+//
+// Style guide: Functions.ModularScale.calculate-modular-spacing
+//
+@function calculate-modular-spacing($step: 0, $breakpoint: 'xl') {
```
Comment (sherakama) [inline core/scss/utilities/functions/modular-scale/_calculate-modular-spacing.scss]: This is great!

File: core/scss/utilities/functions/modular-scale/_get-modular-spacing-base.scss [hunk 19]
```diff
@@ -0,0 +1,15 @@
...
+//
+// Get a value out of the modular-spacing-base map
+// Returns the pixel value of the $modular-spacing-base variable at a breakpoint.
+//
+// $breakpoint - One of `xs`, `sm`, `md`, `lg`, `xl`, `2xl`
+//
+// Style guide: Functions.ModularScale.get-modular-spacing-base
+//
+@function get-modular-spacing-base($breakpoint: 'xs') {
+  @return map-get($modular-spacing-base, $breakpoint);
```
Comment (sherakama) [inline core/scss/utilities/functions/modular-scale/_get-modular-spacing-base.scss]: map-get returns `null` if the key is missing. We should add a warning here if the breakpoint key goes missing. It is likely that someone will modify the grid and breakpoints in the future.

Comment (sherakama) [inline core/scss/utilities/functions/modular-scale/_get-modular-spacing-base.scss]:
```suggestion
  $base: map-get($modular-spacing-base, $breakpoint);
  @if $base == null {
    @warning 'Non-existent key in get-modular-spacing-base: $breakpoint';
  }
  @return $base;
```

File: core/scss/utilities/functions/modular-scale/_get-modular-spacing-multiplier.scss [hunk 20]
```diff
@@ -0,0 +1,20 @@
...
+  @if ( $step >= -4 and $step <= 5 ) {
+    @return map_get($modular-spacing-multiplier, $step);
+  } @else {
```
Comment (sherakama) [inline core/scss/utilities/functions/modular-scale/_get-modular-spacing-multiplier.scss]: Don't need an else statement here as the `@return` won't allow the function to continue.

File: core/scss/utilities/mixins/modular-scale/_modular-spacing.scss [hunk 21]
```diff
@@ -0,0 +1,30 @@
...
+// $property - CSS margin or padding property including directional properties, e.g., margin-left, padding-top etc.
+//
+// **Examples:**
+// - `@include modular-spacing(0, 'margin')`
+// - `@include modular-spacing(-3, 'padding-top')`
+// - `@include modular-spacing(4, 'margin-left');`
+//
+// Style guide: Mixins.Modular.modular-spacing
+//
+@mixin modular-spacing($step: 0, $property: 'padding', $grid: $grid-media) {
```
Comment (sherakama) [inline core/scss/utilities/mixins/modular-scale/_modular-spacing.scss]: 😍 this!

File: core/scss/utilities/mixins/modular-scale/_modular-spacing.scss [hunk 22]
```diff
@@ -0,0 +1,30 @@
...
+  @if ( $step >= -4 and $step <= 5 ) {
```
Comment (sherakama) [inline core/scss/utilities/mixins/modular-scale/_modular-spacing.scss]: Validation is already in the calculate-modular-spacing function. I don't think we need it here.

Comment (yvonnetangsu) [reply core/scss/utilities/mixins/modular-scale/_modular-spacing.scss]: Thanks for the review! I'll go and make those changes 😸

Comment (yvonnetangsu) [reply core/scss/utilities/mixins/modular-scale/_modular-spacing.scss]: Changes have been made including the cc fixes. Let me know if I miss anything else. If it's good to merge, feel free to pull the trigger. Thanks!!

**Pull request merged**